ddtrace/internal/datadog/profiling/include/exporter.hpp

```diff
@@ -0,0 +1,261 @@
+  };
+
+private:
+  unsigned int type_mask;
+  unsigned int max_nframes;
+  unsigned int nframes;
+
+  std::string errmsg;
+
+  // Keeps temporary buffer of frames in the stack
+  std::vector<ddog_prof_Location> locations;
+  std::vector<ddog_prof_Line> lines;
+
+  // Storage for strings
+  std::unordered_set<std::string> strings;
+
+  // Storage for labels
+  ddog_prof_Label labels[8];
+  size_t cur_label = 0;
+
+  // Storage for values
+  std::vector<int64_t> values = {};
+  struct {
+    unsigned short cpu_time;
+    unsigned short cpu_count;
+    unsigned short wall_time;
+    unsigned short wall_count;
+    unsigned short exception_count;
+    unsigned short lock_acquire_time;
+    unsigned short lock_acquire_count;
+    unsigned short lock_release_time;
+    unsigned short lock_release_count;
+    unsigned short alloc_space;
+    unsigned short alloc_count;
+    unsigned short heap_space;
+  } val_idx;
+
+  // Helpers
+  void push_label(const ExportLabelKey key, std::string_view val);
+  void push_label(const ExportLabelKey key, int64_t val);
+  void push_frame_impl(std::string_view name, std::string_view filename,
+                       uint64_t address, int64_t line);
+
+public:
+  uint64_t samples = 0;
+  ddog_prof_Profile *ddog_profile;
+
+  // Clears the current sample without flushing and starts a new one
+  bool start_sample(unsigned int nframes);
+
+  // Add values
+  bool push_walltime(int64_t walltime, int64_t count);
+  bool push_cputime(int64_t cputime, int64_t count);
+  bool push_acquire(int64_t acquire_time, int64_t count);
+  bool push_release(int64_t lock_time, int64_t count);
+  bool push_alloc(uint64_t size, uint64_t count);
+  bool push_heap(uint64_t size);
+
+  // Adds metadata to sample
+  bool push_lock_name(std::string_view lock_name);
+  bool push_threadinfo(int64_t thread_id, int64_t thread_native_id,
+                       std::string_view thread_name);
+  bool push_taskinfo(int64_t task_id, std::string_view task_name);
+  bool push_span_id(int64_t span_id);
+  bool push_local_root_span_id(int64_t local_root_span_id);
+  bool push_trace_type(std::string_view trace_type);
+  bool push_trace_resource_container(std::string_view trace_resource_container);
+  bool push_exceptioninfo(std::string_view exception_type, int64_t count);
+  bool push_class_name(std::string_view class_name);
+
+  // Assumes frames are pushed in leaf-order
+  void push_frame(std::string_view name,     // for ddog_prof_Function
+                  std::string_view filename, // for ddog_prof_Function
+                  uint64_t address,          // for ddog_prof_Location
+                  int64_t line               // for ddog_prof_Line
+  );
+
+  // Flushes the current buffer, clearing it
+  bool flush_sample();
+
+  // Clears temporary things
+  void clear_buffers();
+
+  // Zero out stats
+  void zero_stats();
+
+  bool reset();
+  Profile(ProfileType type, unsigned int _max_nframes = 64);
```

Comment:
Minor: I'm not sure 64 is what you meant ?
Also aligning that with values above might be nice ? I'm not sure this will be a parameter (vs fixed value) if we use it as a storage size for std::arrays.

Reply:
Yeah, overspecifying defaults is kind of hard not to do.

The max_nframes is a user-defined maximum that defines the depth at which unwinding will stop.  It's actually enforced on the Python side, but we use the max_nframes parameter in order to compute how many frames were omitted (since we get the true stack depth).

I think I slightly fixed this in 61f10cc50 by changing parameter defaults and leaning more heavily into the builders.